events/events_network.c

```diff
@@ -233,36 +240,60 @@ events_network_cancel(int s, int op)
 int
 events_network_select(struct timeval * tv)
 {
+	int timeout_ms;
 	size_t i;
 
+	/* Calculate timeout. */
+	if (tv != NULL)
+		timeout_ms = (int)(tv->tv_sec * 1000 + tv->tv_usec * 0.001);
+	else
+		timeout_ms = -1;
+
 	/* Initialize if necessary. */
 	if (initsocketlist())
 		goto err0;
 
-	/* Zero the fd sets... */
-	FD_ZERO(&readfds);
-	FD_ZERO(&writefds);
+	/* Clear list if a rebuild is requested. */
+	if (fds_rebuild_needed) {
+		free(fds);
+		fds = NULL;
+	}
 
-	/* ... and add the ones we care about. */
-	for (i = 0; i < socketlist_getsize(S); i++) {
-		if (socketlist_get(S, i)->reader)
-			FD_SET(i, &readfds);
-		if (socketlist_get(S, i)->writer)
-			FD_SET(i, &writefds);
+	/* Allocate and fill fds list. */
+	if (fds == NULL) {
+		if ((fds = calloc(socketlist_getsize(S),
+		    sizeof(struct pollfd))) == NULL) {
+			warnp("calloc()");
+			goto err0;
+		}
+		nfds = 0;
+		fds_rebuild_needed = 0;
+
+		/* ... and add the ones we care about. */
+		for (i = 0; i < socketlist_getsize(S); i++) {
+			if (socketlist_get(S, i)->reader ||
+			    socketlist_get(S, i)->writer) {
+				fds[nfds].fd = (int)i;
+				if (socketlist_get(S, i)->reader)
```

Comment:
We should check for POLLHUP and POLLERR too -- caused by the arrival of TCP FIN and RST packets, for example -- since these mark the socket as "readable" for the purpose of select (and this behaviour is assumed by users of this code).